Sources/NIOCore/ByteBuffer-lengthPrefix.swift

```diff
@@ -0,0 +1,55 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+extension ByteBuffer {
+    public struct LengthPrefixError: Swift.Error {
+        private enum BaseError {
+            case messageLengthDoesNotFitExactlyIntoRequiredIntegerFormat
+        }
+        private var baseError: BaseError
+        
+        public static let messageLengthDoesNotFitExactlyIntoRequiredIntegerFormat: LengthPrefixError = .init(baseError: .messageLengthDoesNotFitExactlyIntoRequiredIntegerFormat)
+    }
+}
+
+extension ByteBuffer {
+    /// Prefixes a message written by `writeMessage` with the number of bytes written as an `Integer`.
+    /// - Throws: If the number of bytes written during `writeMessage` can not be exactly represented as the given `Integer` i.e. if the number of bytes written is greater than `Integer.max`
+    /// - Returns: Number of total bytes written
+    @discardableResult
+    @inlinable
+    public mutating func writeLengthPrefix<Integer: FixedWidthInteger>(
+        endianness: Endianness = .big,
+        as integer: Integer.Type = Integer.self,
+        message writeMessage: (inout ByteBuffer) throws -> ()
+    ) throws -> Int {
+        let lengthPrefixIndex = writerIndex
+        // Write a zero as a placeholder which will later be overwritten by the actual number of bytes written
+        var totalBytesWritten = writeInteger(.zero, endianness: endianness, as: Integer.self)
+        
+        let messageStartIndex = writerIndex
+        try writeMessage(&self)
+        let messageEndIndex = writerIndex
```

Comment:
I have changed the `writeMessage` function to return a number of bytes written for now. The `set` version of this method actually requires it, otherwise we do not know how many bytes were actually written if only `set` methods are used during `writeMessage`.

However, I find it more convenient to not return the number of bytes written because it can become rather difficult to provide this number if you use any kind of control flow during `writeMessage`. I would like that `setLengthPrefixed` requires a number of bytes written to be returned but `writeLengthPrefixed ` should not require it and compute the bytes written through the indices. WDYT?

As an alternative, we could also require the `writeMessage` function of `writeLengthPrefixed` to return the number of bytes written and add an assert/precondition that this number is actually correct but we should definitely document this behavior.

Reply:
I think I'd rather keep them the same. In principle it shouldn't be too challenging to do this with control flow: a number of NIO projects including NIOSSH effectively include this function and manage their control flow fairly straightforwardly. The pattern tends to be:

```swift
buffer.writeLengthPrefixed(as: UInt32.self) { buffer in
    var written = 0
    written += buffer.writeString(self.someString)
    if self.foo {
       written += buffer.writeInteger(self.bar)
   else {
       written += buffer.writeInteger(self.baz)
    }
    return written
}

```

Reply:
Okay, let's keep it this way.

Just out of curiosity, do we get any benefit compared to the "automatic" calculation ?

Reply:
Yeah, better composition. When the inner function body returns `Void` then you can't write:

```swift
buffer.writeLengthPrefixed(as: UInt32.self) { $0.writeInteger(someInteger) }
```

With this version, you can. It also saves us a bunch of work, computationally. I'm more than happy for us to _police_ the bounds though: we can add an `assert` statement that enforces that the amount the user claims they wrote is actually what they wrote.

Reply:
You can actually write it like in your example because all write methods are marked with `@discardableResult`.

It should also only be two memory reads and one subtraction compared to the manual calculations. 
Maybe my intuitions are wrong but I thought that this should not be noticeable because whatever `writeMessage` does is probably be more expensive.

Reply:
Your intuitions are right, it's not that it's computationally expensive, it's that it's cognitively expensive. The load is simply more complex than allowing the method to say how many bytes it wrote.